server/keyspace/keyspace.go

```diff
@@ -0,0 +1,244 @@
+		}
+		return nil, err
+	}
+
+	return keyspace, nil
+}
+
+// loadKeyspaceID returns the id of keyspace specified by name.
+// It returns error if loading met error, or if keyspace not exists.
+func (manager *Manager) loadKeyspaceID(name string) (uint32, error) {
+	loaded, spaceID, err := manager.store.LoadKeyspaceID(name)
+	if err != nil {
+		return 0, err
+	}
+	if !loaded {
+		return 0, ErrKeyspaceNotFound
+	}
+	return spaceID, nil
+}
+
+// LoadKeyspace returns the keyspace specified by name.
+// It returns error if loading or unmarshalling met error or if keyspace does not exist.
+func (manager *Manager) LoadKeyspace(name string) (*keyspacepb.KeyspaceMeta, error) {
+	spaceID, err := manager.loadKeyspaceID(name)
+	if err != nil {
+		return nil, err
+	}
+	keyspace := &keyspacepb.KeyspaceMeta{}
+	loaded, err := manager.store.LoadKeyspace(spaceID, keyspace)
+	if err != nil {
+		return nil, err
+	}
+	if !loaded {
+		return nil, ErrKeyspaceNotFound
+	}
+	return keyspace, nil
+}
+
+// UpdateKeyspace update keyspace's state and config according to request.
+// It returns error if saving failed, operation not allowed, or if keyspace not exists.
+func (manager *Manager) UpdateKeyspace(request *UpdateKeyspaceRequest) (*keyspacepb.KeyspaceMeta, error) {
+	manager.metaLock.Lock()
+	defer manager.metaLock.Unlock()
+	// load keyspace by id
+	keyspace, err := manager.LoadKeyspace(request.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// disallow modifying archived keyspace
+	if keyspace.State == keyspacepb.KeyspaceState_ARCHIVED {
+		return nil, errKeyspaceArchived
+	}
+
+	// update keyspace state and config
+	if request.UpdateState {
+		if err = changeKeyspaceState(keyspace, request.NewState, request.Now); err != nil {
+			return nil, err
+		}
+	}
+	changeKeyspaceConfig(keyspace, request.ToPut, request.ToDelete)
+	// save keyspace
+	if err = manager.store.SaveKeyspace(keyspace); err != nil {
+		return nil, err
+	}
+	return keyspace, nil
+}
+
+// LoadRangeKeyspace load up to limit keyspaces starting from keyspace with startID.
+func (manager *Manager) LoadRangeKeyspace(startID uint32, limit int) ([]*keyspacepb.KeyspaceMeta, error) {
+	return manager.store.LoadRangeKeyspace(startID, limit)
+}
+
+// changeKeyspaceState changes the state of given keyspace meta.
+// It also records the time state change happened.
+func changeKeyspaceState(keyspace *keyspacepb.KeyspaceMeta, newState keyspacepb.KeyspaceState, now time.Time) error {
+	if keyspace.State == newState {
+		return nil
+	}
+	// ARCHIVED is the terminal state that cannot be changed from.
+	if keyspace.State == keyspacepb.KeyspaceState_ARCHIVED {
+		return errKeyspaceArchived
+	}
+
+	keyspace.StateChangedAt = now.Unix()
+	keyspace.State = newState
+	return nil
+}
+
+// changeKeyspaceConfig update a keyspace's config according to toPut and toDelete.
+func changeKeyspaceConfig(keyspace *keyspacepb.KeyspaceMeta, toPut map[string]string, toDelete []string) {
```

Comment:
If we implement delete here, it is possible that the same key is updated first and then deleted during the update process.So it will increase the complexity of API.

Divide update and delete into different interfaces is better?